spring-integration-core/src/test/java/org/springframework/integration/router/HeaderValueRouterTests.java

```diff
@@ -183,7 +195,33 @@ public void resolveMultipleChannelsWithCommaDelimitedString() {
 		assertNotNull(result2);
 		assertSame(message, result1);
 		assertSame(message, result2);
+		context.close();
 	}
 
+	@Test
+	public void dynamicChannelCache() {
+		StaticApplicationContext context = new StaticApplicationContext();
+		RootBeanDefinition routerBeanDefinition = new RootBeanDefinition(HeaderValueRouter.class);
+		routerBeanDefinition.getConstructorArgumentValues().addGenericArgumentValue("testHeaderName");
+		routerBeanDefinition.getPropertyValues().addPropertyValue("resolutionRequired", "true");
+		routerBeanDefinition.getPropertyValues().addPropertyValue("dynamicChannelLimit", "2");
+		context.registerBeanDefinition("router", routerBeanDefinition);
+		context.registerBeanDefinition("channel1", new RootBeanDefinition(QueueChannel.class));
+		context.registerBeanDefinition("channel2", new RootBeanDefinition(QueueChannel.class));
+		context.registerBeanDefinition("channel3", new RootBeanDefinition(QueueChannel.class));
+		context.refresh();
+		MessageHandler handler = (MessageHandler) context.getBean("router");
+		String channels = "channel1, channel2, channel1, channel3";
+		Message<?> message = MessageBuilder.withPayload("test").setHeader("testHeaderName", channels).build();
+		handler.handleMessage(message);
+		QueueChannel channel1 = (QueueChannel) context.getBean("channel1");
+		QueueChannel channel2 = (QueueChannel) context.getBean("channel2");
+		QueueChannel channel3 = (QueueChannel) context.getBean("channel3");
+		assertThat(channel1.getQueueSize(), equalTo(2));
+		assertThat(channel2.getQueueSize(), equalTo(1));
+		assertThat(channel3.getQueueSize(), equalTo(1));
+		assertThat(context.getBean(HeaderValueRouter.class).getDynamicChannelNames(), contains("channel1", "channel3"));
+		context.close();
```

Comment:
Fully looks like this test can dispense with context and be more cleaner.
But it is minor, of course.
All test in this class are similar :smile: